TLCALIPER-115 - Reconcile events

lsloan opened the pull request:
JIRA: https://itsjira.umms.med.umich.edu/browse/TLCALIPER-115
Branch: https://github.com/lsloan/caliper-event-analysis/tree/TLCALIPER-115

See `README.md` for documentation.  Following its steps or using the included scripts will show the matches and differences between input files of log and event data.

It runs so fast (even on 1.9 million lines of data) and is so simple that it feels like something may be wrong with it.  Tell me if you find any holes in the logic.

lsloan committed:
CALIPER-115 - Add validation code and documentation.

lsloan committed:
TLCALIPER-115 - Initial documentation

lsloan committed:
TLCALIPER-115 - "reconcile" is a more accurate term than "validate" for this process.

lsloan committed:
CALIPER-115 - Initial version of script based on documentation

lsloan committed:
TLCALIPER-115 - Script of jq filter to get important properties from Caliper JSONL input and produce TSV output.

lsloan committed:
TLCALIPER-115 - Updated documentation and scripts.

gsilver commented on src/reconcile/caliperEventTsv.sh:
```diff
@@ -0,0 +1,6 @@
+# Send input through stdin or give a filename as an argument
+
+jq -r '[.eventTime[:19]+"Z", .actor.name, .object["@id"], .generated.value//"0"]|@tsv' ${1:--} | \
+    sort | sed 's/%40/@/g'
```
This works blazingly fast.  I like the idea of paring down the events to Completed and Skipped via grep first. We are not interested in reconciling other types of events?

lsloan replied:
The PR app. log (stored in its MySQL DB) contains only information analogous to Caliper events with the AssessmentItemEvent type and the Completed or Skipped actions.  According to @pushyamig, the log doesn't contain information about other Caliper event types that the application is emitting, like AnnotationEvent, AssessmentEvent, NavigationEvent, or SessionEvent.

Reducing the dataset using grep is one of the advantages of working with a JSONL file.  Rather than a file containing valid JSON end-to-end, we have a file with one object per line.  It helps to select only a few objects (millions) out of many (millions more).

gsilver reviewed (commented):
Could not do end to end testing as I have no access to the log data. I limited myself to testing the formatting of Caliper data via grep and jq.

dlhaines commented:
Looks good overall.

If the prior version of the tsv script had a pipe, or needed to start a shell for each line for some other reason, then the big time difference would be plausible.

I'd add script start / end timestamps to the output.

dlhaines reviewed (commented):
Looks good overall.

If the prior version of the tsv script had a pipe, or needed to start a shell for each line for some other reason, then the big time difference would be plausible.

I'd add script start / end timestamps to the output.

lsloan committed:
TLCALIPER-115 - Keep only unique event data

lsloan committed:
TLCALIPER-115 - Use a temp. file

By using a temporary file, we can monitor the activity and progress of the program.

lsloan committed:
TLCALIPER-115 - Aggregate results by date

lsloan committed:
TLCALIPER-115 - Add timestamps and date file

Add start/end timestamp messages.  Make a file of all dates found in the output files.

lsloan committed:
TLCALIPER-115 - Minor aggregation corrections
